### PR TITLE
Enforce a namespace policy.

### LINKED ISSUE #341: Constrain some namespaces to only allow certain branches
This was first reported downstream in the modularity efforts here: http://taiga.fedorainfracloud.org/project/modularity/us/130?kanban-status=470

We can maybe do this with a simple config file entry that sets a 'branch policy' for certain namespaces.. something that just forbids creating branches on this or that namespace if the branchname matches a given regex?

### PR DESCRIPTION
**Pull request by ralphbean**
Fixes #341.

I think this is correct.  Did I miss any other API endpoints that need this?

### CODE REVIEW

**Comment by pypingou** [comment]
> Did I miss any other API endpoints that need this?

I didn't check the code, but don't the UI endpoints need to be adjusted as well?

Also, the mass_branching script needs to be tweaked as otherwise it will create the branches for these namespaces as well

**Comment by pypingou** [comment]
Also, we should consider adding a few tests for these (either in the current suite or in a new file)

**Comment by ralphbean** [comment]
> don't the UI endpoints need to be adjusted as well?

How so?

**Comment by ralphbean** [comment]
> the mass_branching script needs to be tweaked

Where does that script live?  I couldn't find it in the repo.

**Comment by pypingou** [comment]
>>   don't the UI endpoints need to be adjusted as well?
> How so?

Like the list of branches when an user requests a new one?

**Comment by pypingou** [comment]
>> the mass_branching script needs to be tweaked

> Where does that script live? I couldn't find it in the repo.

The script is at:
https://github.com/fedora-infra/pkgdb2/blob/master/utility/pkgdb2_branch.py
but its logic is at:
https://github.com/fedora-infra/pkgdb2/blob/master/pkgdb2/lib/__init__.py#L1684-L1782

**Comment by ralphbean** [comment]
OK - that should address all the comments so far.

File: pkgdb2/lib/__init__.py [hunk 1]
```diff
@@ -1730,7 +1739,8 @@ def add_branch(session, clt_from, clt_to, user):
     SELECT package_id, point_of_contact, %s, status, critpath, '%s'
     FROM "PackageListing"
     WHERE "PackageListing".collection_id = %s
-    AND "PackageListing".status IN ('Approved','Orphaned');''' % (
+    AND "PackageListing".status IN ('Approved','Orphaned')
```
Comment (pypingou) [inline pkgdb2/lib/__init__.py]: we lost the ';' not sure it matters though

File: pkgdb2/lib/__init__.py [hunk 2]
```diff
@@ -1743,9 +1753,16 @@ def add_branch(session, clt_from, clt_to, user):
...
     AND p1.package_id = p2.package_id
-    AND "PackageListingAcl".packagelisting_id = p1.id;
+    AND "PackageListingAcl".packagelisting_id = p1.id
     ''' % (datetime.utcnow(), clt_from.id, clt_to.id)
 
+    for namespace in exempted_namespaces:
+        q1 += ''' AND "PackageListing".namespace != '%s' ''' % namespace
+        q2 += ''' AND p1.namespace != '%s' ''' % namespace
+
+    q1 += ";"
```
Comment (pypingou) [inline pkgdb2/lib/__init__.py]: And there it is :)

**Comment by pypingou** [comment]
Jenkins seems unhappy:
```
  File "<http://jenkins.fedorainfracloud.org/job/PackageDB2/ws/pkgdb2/ui/packages.py",> line 1293, in package_request_
+branch
    policy = namespace_policy.get(pkg.namespace)
AttributeError: 'PackageListing' object has no attribute 'namespace'
```

**Comment by pypingou** [comment]
The rest looks fine to me :)

**Comment by ralphbean** [comment]
Ugh, this also means that for ``AND "PackageListing".namespace`` that `namespace` is not a thing there either.  It is associated with the ``Package`` which is associated with the ``PackageListing``.  It is not an attribute of the listing itself.

**Comment by ralphbean** [comment]
- eabacf7 should make jenkins happy.
- 48f3c50 requires a close look.  I don't know how to test it.  :/

**Comment by pypingou** [comment]
>  I don't know how to test it. :/

The way I tested it is by running a local pkgdb, with a dump from prod and doing branching in it :]

**Comment by pypingou** [comment]
````
$ time PKGDB2_CONFIG=../config python utility/pkgdb2_branch.py --user=pingou --groups=sysadmin-main --nomail f26
Updating rawhide's dist-tag
New dist-tag: `.fc27` [Y/n]: 
Create the new collection
  name:          Fedora
  version:       26
  status:        Under Development
  branchname:    f26
  disttag:       .fc26
  koji:          f26
  allow_retire:  True
Save this new collection? [Y/n]: 
Branching for f26...
Output from the branching:

SUCCESS: successfully branched (PackageListing) Fedora from to Fedora 26
SUCCESS: successfully branched (PackageListingAcl) Fedora from to Fedora 26
PKGDB2_CONFIG=../config python utility/pkgdb2_branch.py --user=pingou   f26  0.61s user 0.07s system 0% cpu 1:48:33.38 total
````

I've updated the mass-branch script so that it can update master and create the new collection just before we do the branching. But at least, on my local copy of pkgdb2 the branching worked fine for regular branches/namespaces (I didn't add a `modules` namespace for testing, I can do that still).

**Comment by pypingou** [comment]
Ok, so I added a ``modules`` namespace, added a couple of package to it, the ran the ``pkgdb2_branch.py`` script, and :

````
    modules/foo -- bar ( master f27 )
    modules/test_module -- new module ( master f27 f26 )
````
It branched :(

File: tests/test_pkgdblib.py [hunk 3]
```diff
@@ -1684,6 +1684,14 @@ def test_add_branch(self):
         self.assertEqual(pkg_acl[2].collection.branchname, 'f19')
         self.assertEqual(len(pkg_acl[2].acls), 5)
 
+        # Check that the namespace policy took effect
```
Comment (pypingou) [inline tests/test_pkgdblib.py]: Should we test it before as well?

Comment (ralphbean) [inline tests/test_pkgdblib.py]: Done in f1e1f74.

**Comment by pypingou** [comment]
Cool, thanks :)

:+1: to merge :)

**Comment by ralphbean** [comment]
Thanks for working through it with me, @pypingou!

**Pull request merged**